### PR TITLE
Bc refactor

### PR DESCRIPTION
**Pull request by charlieowen3065**
For this PR, I am trying to add my changes to the way that boundary conditions (BCs) are handled. A new BC class was added for a general Dirichlet BC. This class is able to handle:

1. Mass flow rate
2. Pressure
3. Temperature
4. Enthalpy

For now, these need to be hard-coded in, though some thought should be put into streamlining this process. A better system would be in connecting the BCs to the variables/data containers themselves, though this will have to wait for its own issue.

The system class was also adjusted to account for these changes to BCs. For now, the old inputs are kept, though will need to be removed later and replaced with only the new method of BCs.

This adjustment mainly changes the way that BC are made in the input, which now ties more tightly the input and the variable they are acting upon (where before, the BC was closely tied to the _equations_ that it acted upon, which seems less intuitive and will make generalization harder). Now, the input for a BC will look more like:

```
"boundary_conditions" : {
    "mass_flow_rate" : {"inlet" : 2.5}
    "pressure" : {"outlet" : 1e5}
    "temperature" : {"inlet" : 700}
}
```

This PR will also be closely tied to a PR for SyTH.

### CODE REVIEW

**Comment by charlieowen3065** [comment]
I had forgotten to test the "tests" in flowforge, and have only tested those in syth. 

Let me work on those and make those updates soon.

**Comment by charlieowen3065** [comment]
@cgentry7 

I've added a slight modification to how the BCs are input and handled, which I think will make the process more intuitive here and simpler to add additional BC types.

Now, the input file will look more similar to:
```
bounday_conditions = {
        "unique_boundary_name" : {"boundary_type": "DirichletBC", "surface": "surface_name", "varaible": "varaible_name",  "value", float},
        "inlet_mdot"           : {"boundary_type": "DirichletBC", "surface": "inlet",        "varaible": "mass_flow_rate", "value", 25.0},
        "outlet_pressure"      : {"boundary_type": "DirichletBC", "surface": "outlet",       "varaible": "pressure",       "value", 1e5},
        "inlet_temperature"    : {"boundary_type": "DirichletBC", "surface": "inlet",        "varaible": "temperature",    "value", 700}
```

where each input BC is given a unique name, which allows for multiple temperature, pressure, etc BCs to be specified. Within each BC object, the boundary type, associated variable name, and the input value are given.

By asking for the boundary type, this allows for the easy implementation of Neumann/Robin BCs in the future, and all that will be needed to add is a new class for them.

I also created a GeneralBC class that DirichletBC inherits from, and which Neumann or Robin BCs could inherit off from.

FlowForge then passes the BoundaryConditions class which holds all the BC objects to syth.

I think this is a better way to do this, and now I just need to update syth to account for these changes, which is what I will do today and tomorrow.

File: flowforge/input/BoundaryConditions.py [hunk 1]
```diff
@@ -112,17 +116,110 @@ def type_outlet(self):
...
+      - enthalpy
+
+    bounday_conditions should be defined as a dict in the form:
+
+    bounday_conditions = {
+        "mass_flow_rate" : {"inlet/outlet", float},
+        "pressure"       : {"inlet/outlet", float},
+        "temperature"    : {"inlet/outlet", float},
+        "enthalpy"       : {"inlet/outlet", float}
+    }
```
[the remark this answers is not in the document]
Comment (cgentry7) [reply flowforge/input/BoundaryConditions.py]: Addressed by revisions

File: flowforge/input/BoundaryConditions.py [hunk 2]
```diff
@@ -112,17 +116,110 @@ def type_outlet(self):
...
+        - Perhaps : _temperature -> dict : {"node1" : float, "node2" : float, ...}
+        - Need to examine how this would function downstream before adding too much
+
+      - Also, as a more long-term note, this class is *not* in the final form. As a goal, this
+        class should not have to specify/hard-code the properties (temperature, mdot, ...), but
+        should instead strive to be more general. It should take in, as inputs, a reference to
+        the variable of which the BC will be applied, the value at the boundary, and some mapping
+        to the actual fluid mesh, as well as the mesh-index at which to apply the BC. In this
+        manner, one could theortically set a BC of any variable at any point in the mesh. As this
+        is not a vitale feature of SyTH, it should be done later as an added feature.
```
[the remark this answers is not in the document]
Comment (cgentry7) [reply flowforge/input/BoundaryConditions.py]: Addressed by revisions

File: flowforge/input/BoundaryConditions.py [hunk 3]
```diff
@@ -112,17 +116,140 @@ def type_outlet(self):
...
+        bc_objects = {"DirichletBC": DirichletBC}
+
+        bc_list = []
+        for bc_name in bc_names:
+            bc_type = bounday_conditions[bc_name]["boundary_type"]
+            bc_obj = bc_objects[bc_type]
+            bc = bc_obj(bounday_conditions["surface"],
+                        bounday_conditions["variable"],
+                        bounday_conditions["value"])
+            bc_list.append(bc)
```
Comment (cgentry7) [inline flowforge/input/BoundaryConditions.py:164]: I think you're missing a `[bc_name]` when retrieving `surface`, `variable`, and `value` off the `boundary_conditions` dictionary.

Actually, I'm wondering why you didn't do something like this:

```
self.bcs = {}
for bc_name, bc in bound_conditions.items():
    bc_type = bc["boundary_type"]
    bc_obj = bc_objects[bc_type]
    self.bcs[bc_name] = bc_obj(bc["surface"], bc["variable"], bc["value"]))
```

File: flowforge/input/BoundaryConditions.py [hunk 4]
```diff
@@ -112,17 +116,140 @@ def type_outlet(self):
...
+    @property
+    def boundary_value(self):
+        return self._value
+
+    @boundary_value.setter
+    def boundary_value(self, value):
+        self._value = value
+
+    @property
+    def varaible_name(self):
```
Comment (cgentry7) [inline flowforge/input/BoundaryConditions.py:225]: "variable" typo

File: flowforge/input/BoundaryConditions.py [hunk 5]
```diff
@@ -112,17 +116,140 @@ def type_outlet(self):
...
+
+    def _get_variable_conversion(self, uc: UnitConverter):
+        if self.varaible_name == "mass_flow_rate":
+            conversion_factor = uc.massFlowRateConversion
+        elif self.varaible_name == "pressure":
+            conversion_factor = uc.pressureConversion
+        elif self.varaible_name == "temperature":
+            conversion_factor = uc.temperatureConversion(self.boundary_value) / self.boundary_value
+        elif self.varaible_name == "enthalpy":
+            conversion_factor = uc.enthalpyConversion
```
Comment (cgentry7) [inline flowforge/input/BoundaryConditions.py:244]: "variable" typo

**Review by cgentry7 — CHANGES_REQUESTED** [review]
Minor comments, otherwise looks good.

File: flowforge/input/BoundaryConditions.py [hunk 6]
```diff
@@ -112,17 +116,110 @@ def type_outlet(self):
...
+        class should not have to specify/hard-code the properties (temperature, mdot, ...), but
+        should instead strive to be more general. It should take in, as inputs, a reference to
+        the variable of which the BC will be applied, the value at the boundary, and some mapping
+        to the actual fluid mesh, as well as the mesh-index at which to apply the BC. In this
+        manner, one could theortically set a BC of any variable at any point in the mesh. As this
+        is not a vitale feature of SyTH, it should be done later as an added feature.
+                                                                                        - Charlie
+    """
+
+    def __init__(self, **bounday_conditions: dict):
```
[the remark this answers is not in the document]
Comment (cgentry7) [reply flowforge/input/BoundaryConditions.py]: Addressed by revisions

File: flowforge/input/BoundaryConditions.py (hunk 3 above)
Comment (charlieowen3065) [reply flowforge/input/BoundaryConditions.py:164]: I have gone back through and replaced it with this. This is a much more compact way to do exactly what I wanted - thanks!

I'll also go through my syth code and make sure I don't have any similar issues, and generally clean that up once this PR is complete.

**Review by cgentry7 — APPROVED** [review]

File: flowforge/input/BoundaryConditions.py [hunk 7]
```diff
@@ -112,17 +116,131 @@ def type_outlet(self):
...
-            elif self.type_outlet == 'enthalpy':
+            elif self.type_outlet == "enthalpy":
                 self._val_outlet *= uc.enthalpyConversion
             else:
                 raise Exception("Unknown enthalpy BC type: " + self.type_outlet)
+
+
+class BoundaryConditions:
+    """
+    Continer class for all input boundary conditions
```
Comment (MicahBest) [inline flowforge/input/BoundaryConditions.py:137]: "Container"

File: flowforge/input/BoundaryConditions.py [hunk 8]
```diff
@@ -112,17 +116,131 @@ def type_outlet(self):
...
+
+    "bounday_conditions" should be defined as a dict in the form:
+
+    bounday_conditions = {
+        "unique_boundary_name" :
+                {"boundary_type": "DirichletBC", "surface": "surface_name", "varaible": "varaible_name",  "value", float},
```
Comment (MicahBest) [inline flowforge/input/BoundaryConditions.py:143]: "variable", "variable_name"

File: flowforge/input/BoundaryConditions.py [hunk 9]
```diff
@@ -112,17 +116,131 @@ def type_outlet(self):
...
+        "inlet_mdot"           :
+                {"boundary_type": "DirichletBC", "surface": "inlet",        "varaible": "mass_flow_rate", "value", 25.0},
```
Comment (MicahBest) [inline flowforge/input/BoundaryConditions.py:145]: "variable"

File: flowforge/input/BoundaryConditions.py [hunk 10]
```diff
@@ -112,17 +116,131 @@ def type_outlet(self):
...
+        "outlet_pressure"      :
+                {"boundary_type": "DirichletBC", "surface": "outlet",       "varaible": "pressure",       "value", 1e5},
```
Comment (MicahBest) [inline flowforge/input/BoundaryConditions.py:147]: "variable"

File: flowforge/input/BoundaryConditions.py [hunk 11]
```diff
@@ -112,17 +116,131 @@ def type_outlet(self):
...
+        "inlet_temperature"    :
+                {"boundary_type": "DirichletBC", "surface": "inlet",        "varaible": "temperature",    "value", 700}
```
Comment (MicahBest) [inline flowforge/input/BoundaryConditions.py:149]: "variable"

File: flowforge/input/BoundaryConditions.py [hunk 12]
```diff
@@ -112,17 +116,131 @@ def type_outlet(self):
...
+    }
+    """
+    def __init__(self, **bounday_conditions: dict):
```
Comment (MicahBest) [inline flowforge/input/BoundaryConditions.py:152]: "**boundary_conditions"

File: flowforge/input/BoundaryConditions.py [hunk 13]
```diff
@@ -112,17 +116,131 @@ def type_outlet(self):
...
+
+        bc_objects = {"DirichletBC": DirichletBC}
+
+        self.bcs = {}
+        for bc_name, bc in bounday_conditions.items():
```
Comment (MicahBest) [inline flowforge/input/BoundaryConditions.py:157]: "boundary_conditions"

**Review by MicahBest — CHANGES_REQUESTED** [review]
I know I wasn't requested to review this, but I couldn't help but notice a few spelling errors so I've made a few comments.

**Comment by charlieowen3065** [comment]
@MicahBest 

Thanks Micah! Just went through and fixed those spelling mistakes.

**Review by MicahBest — APPROVED** [review]

**Comment by MicahBest** [comment]
No problem. Looks good.

**Pull request merged**